foxai/explainer/computer_vision/algorithm/xrai.py

```diff
@@ -0,0 +1,687 @@
+                        original_shape,
+                        order=0,
+                        preserve_range=True,
+                        mode="constant",
+                        anti_aliasing=False,
+                    ).astype(int)
+                segment_list.append(segment)
+
+        masks = _unpack_segs_to_masks(segment_list)
+        if dilation_rad:
+            masks = apply_dilation(dilation_rad=dilation_rad, masks=masks)
+
+        batch_masks.append(masks)
+    return batch_masks
+
+
+def apply_dilation(dilation_rad: int, masks: List[np.ndarray]) -> List[np.ndarray]:
+    """Apply morphological dilation of images.
+
+    Args:
+        dilation_rad: Sets how much each segment is dilated to include edges,
+            larger values cause more blobby segments, smaller values
+            get sharper areas.
+        masks: List of masks to process.
+
+    Returns:
+        List of dilated masks.
+    """
+    footprint = disk(dilation_rad)
+    return [dilation(mask, footprint=footprint) for mask in masks]
+
+
+def _gain_density(
+    mask1: np.ndarray, attributes: np.ndarray, mask2: Optional[np.ndarray] = None
+) -> float:
+    """Calculate gain density as mean value in attributes mask.
+
+    Args:
+        mask1: Compute density for mask1.
+        attributes: Attributes array used to compute gain density.
+        mask2: If mask2 is specified, compute density
+            for mask1 \\ mask2. Defaults to None.
+
+    Returns:
+        Gain density score.
+    """
+    if mask2 is None:
+        added_mask = mask1
+    else:
+        added_mask = _get_diff_mask(mask1, mask2)
+    if not np.any(added_mask):
+        return -np.inf
+    else:
+        return attributes[added_mask].mean()
+
+
+def _get_diff_mask(array1: np.ndarray, array2: np.ndarray) -> np.ndarray:
+    """Compute element-wise AND operation between two arrays.
+
+    Args:
+        array1: First array.
+        array2: Second array.
+
+    Returns:
+        Array of booleans representing result of element-wise AND operation
+            between arrays.
+    """
+    return np.logical_and(array1, np.logical_not(array2))
+
+
+def _get_diff_cnt(array1: np.ndarray, array2: np.ndarray) -> int:
+    """Compute sum of different elements in 2 arrays.
+
+    Args:
+        array1: First array.
+        array2: Second array.
+
+    Returns:
+        Number of different elements between two arrays.
+    """
+    return np.sum(_get_diff_mask(array1, array2))
+
+
+def _unpack_segs_to_masks(segment_list: List[np.ndarray]) -> List[np.ndarray]:
+    """Create masks from segments.
+
+    Args:
+        segment_list: List of segments.
+
+    Returns:
+        List of masks.
+    """
+    masks: List[np.ndarray] = []
+    for segment in segment_list:
+        for index in range(segment.min(), segment.max() + 1):
+            masks.append(segment == index)
+    return masks
+
+
+class XRAI:
+    """A CoreSaliency class that computes saliency masks using the XRAI method."""
```

Comment:
This docstring seems off.

Reply:
Fixed: https://github.com/softwaremill/FoXAI/pull/99/commits/49a269c31fe09c365852daba10e1a93d5fe9850a.